README.md
# madhack_vrx

## About

The MadHack VRX Project is an DC Augmented Reality project to resolve Data Centers Alarms Problems maintained by Oracle EMEA.

![VRX Ideas](https://github.com/operard/madhack_vrx/blob/master/vrx_ideas.png)

## Home page 

The project documentation is [here](/doc/README.md).

You can find all tutorials for VRX Project:

- <a href="https://htmlpreview.github.io/?https://github.com/operard/madhack_vrx/blob/master/functions/vrx_fn.htm">Functions Tutorial</a>
- [Splunk Tutorial](/splunk/README.md)
- [Crontab Tutorial](/crontab/README.md)

## Download and Installation

The VRX Solution can be downloaded from the [releases section on GitHub](https://github.com/operard/madhack_vrx).

- <a href="https://htmlpreview.github.io/?https://github.com/operard/madhack_vrx/blob/master/functions/vrx_fn.htm">Functions Tutorial</a>
- [Splunk Tutorial](/splunk/README.md)
- [Crontab Tutorial](/crontab/README.md)

## Changes

See [CHANGELOG](/CHANGELOG.md).

## Contributing

To contribute, see [CONTRIBUTING](/CONTRIBUTING.md) for details.

##### Deploying the VRX Solution

After you clone the repository on GitHub, you will need to deploy the VRX Solution.

1. Install Splunk Docker Container.
2. Deploy Splunk Cloud Gateway.
3. Deploy Functions Solutions.
4. Deploy Crontab scripts or triggers functions Fn.

## Blog Announcements


## License

Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.

Licensed under the Universal Permissive License v 1.0 as shown at http://oss.oracle.com/licenses/upl




CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## 1.0.0
### Added
- Initial Release
- Support added for Oracle Cloud Infraastructure OCI and Splunk Enterprise Solution.


functions/func.py
import io
import sys
import json
import oci
import oci.monitoring as monitoring
import oci.monitoring.models as models
from datetime import datetime, timedelta
from fdk import response

class OCIGetMetrics:
    def __init__(self, compartment_id, name_space, query):
        self._compartmentID = compartment_id
        self._name_space = name_space
        self._query = query
        self.monitoring_client = self.init_monitor_client()

    @property
    def compartment_id(self):
        return self._compartmentID

    @property
    def name_space(self):
        return self._name_space

    @property
    def query(self):
        return self._query

    @staticmethod
    def init_monitor_client():
        # Load the default configuration
        # config = oci.config.from_file()

        monitoring_client = monitoring.MonitoringClient(config)
        return monitoring_client

    def get_metrics(self):
        end_time = datetime.utcnow()
        start_time = str((datetime.utcnow() - timedelta(hours=1, minutes=5)).isoformat()[:-3]) + 'Z'

        summarize_metrics_data_details = models.SummarizeMetricsDataDetails(
            namespace=self.name_space,
            query=self.query,
            start_time=start_time,
            end_time=end_time
        )
        response = self.monitoring_client.summarize_metrics_data(
            compartment_id=self.compartment_id,
            summarize_metrics_data_details=summarize_metrics_data_details)
        return response



def handler(ctx, data: io.BytesIO=None):
    compartmentID = "ocid1.compartment.oc1..aaaaaaaa4gk5fmtbrfnkfrwwkoef5pmrvxu7dauh52hbisvnhx6rgas5jxja"
    namespace = "oci_computeagent"
    query = "CPUUtilization[1m]{resourceId = \"ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva\"}.mean()"
    try:
        body = json.loads(data.getvalue())
        # query = body.get("query")
        name = body.get("name")
        metrics_obj = OCIGetMetrics(compartmentID, namespace, query)
        if sys.getsizeof(name) > 40:
           name = "Toni"
        databeta = metrics_obj.get_metrics().data

    except (Exception, ValueError) as ex:
        print(str(ex))

    return response.Response(
        ctx, response_data=json.dumps(
            {"availabilityDomain": "VrTN:EU-FRANKFURT-1-AD-3", 
             "cpu_mean": 6.094046517476844, 
             "faultDomain": "FAULT-DOMAIN-3", 
             "host": "Ubuntu-madhack-vrx", 
             "mem_mean": 3.931413040783038, 
             "net_in": 18619.02, 
             "net_out": 20041.54, 
             "region": "eu-frankfurt-1", 
             "shape": "VM.Standard2.4", 
             "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S")}),
          #    {"message": "hola {0}".format(name)}),
        headers={"Content-Type": "application/json"}
    )
    # return metrics_obj.get_metrics().data


crontab/oci_metrics_vrx_fn.sh
#!/bin/bash
#************************************************************************
#
#   oci_metrics_vrx-fn.sh - get OCI Metrics using APIGW & Functions
#   metadata information into JSON files.
#
#   Copyright 2019  Olivier Perard 
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
#************************************************************************
# Available at: https://github.com/operard/madhack_vrx/blob/master/crontab/
# Created on: 03/11/2019 by Olivier Perard
# Version 1.00
#************************************************************************
set -e

# Define paths for oci-cli and jq or put them on $PATH. Don't use relative PATHs in the variables below.
v_oci="oci"
v_jq="jq"

apigw=$(curl -k -X GET https://cvp4qdwzed5hxig7cfcb73utde.apigateway.eu-frankfurt-1.oci.customer-oci.com/v1/getocimetrics)

echo "$apigw" >> /tmp/oci_metrics_vrx_fn.log

curl -k "http://130.61.28.47:18088/services/collector" \
    -H "Authorization: Splunk b81fe15a-0be4-4a1e-82d5-5324888d78f0" \
	-d '{"sourcetype": "_json", "event": '$apigw'}'



functions/java/MonitoringVrxFunction.java
package io.fnproject.monitoring;

import com.oracle.bmc.auth.ResourcePrincipalAuthenticationDetailsProvider;

import com.oracle.bmc.monitoring.MonitoringClient;
import com.oracle.bmc.monitoring.model.ListMetricsDetails;
import com.oracle.bmc.monitoring.requests.ListMetricsRequest;
import com.oracle.bmc.monitoring.responses.ListMetricsResponse;

import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;
import java.util.stream.Collectors;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.cloudevents.CloudEvent;

import java.io.IOException;
import java.io.InputStream;
import java.text.SimpleDateFormat;

public class MonitoringVrxFunction {

    private MonitoringClient monitoringClient = null;

    final ResourcePrincipalAuthenticationDetailsProvider provider
            = ResourcePrincipalAuthenticationDetailsProvider.builder().build();

    public MonitoringVrxFunction() {
        try {

            //print env vars in Functions container
            System.err.println("OCI_RESOURCE_PRINCIPAL_VERSION " + System.getenv("OCI_RESOURCE_PRINCIPAL_VERSION"));
            System.err.println("OCI_RESOURCE_PRINCIPAL_REGION " + System.getenv("OCI_RESOURCE_PRINCIPAL_REGION"));
            System.err.println("OCI_RESOURCE_PRINCIPAL_RPST " + System.getenv("OCI_RESOURCE_PRINCIPAL_RPST"));
            System.err.println("OCI_RESOURCE_PRINCIPAL_PRIVATE_PEM " + System.getenv("OCI_RESOURCE_PRINCIPAL_PRIVATE_PEM"));

            monitoringClient = new MonitoringClient(provider);

        } catch (Throwable ex) {
            System.err.println("Failed to instantiate MonitoringClient client - " + ex.getMessage());
        }
    }

    public String getStatus(String input){
        if (monitoringClient == null) {
            System.err.println("There was a problem creating the Monitoring Client object. Please check logs");
            return "";
        }
        // Generate Date in ISO 8601
        Date date = new Date(System.currentTimeMillis());
	    // Conversion
	    SimpleDateFormat sdf;
	    sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
	    sdf.setTimeZone(TimeZone.getTimeZone("CET"));
	    String ttime = sdf.format(date);

	 // {"timestamp": '$timestamp', "host": '$host', "region": '$region', "availabilityDomain": '$availabilityDomain', 
	 // "faultDomain": '$faultDomain', "shape": '$shape', 
	 // "cpu_mean": '$cpuval', "mem_mean": '$memval', "net_in": '$netinval', "net_out": '$netoutval' }

		String compartment = "ocid1.compartment.oc1..aaaaaaaa4gk5fmtbrfnkfrwwkoef5pmrvxu7dauh52hbisvnhx6rgas5jxja";
		String namespace = "oci_computeagent";
		
		String host = "Ubuntu-madhack-vrx";
		String region ="eu-frankfurt-1";
		String availabilityDomain = "VrTN:EU-FRANKFURT-1-AD-3";
		String faultDomain = "FAULT-DOMAIN-3";
		String shape = "VM.Standard2.4";
		
		
		String metricName;
		ListMetricsResponse resp;
		
		metricName = "CPUUtilization[1m]{resourceId = \"ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva\"}.mean()";
		resp= executeQuery(compartment, namespace, metricName);
		// response.getItems(); 
		// 4.214563820932875|3.850824792171347|2126.26|1966.16
		String cpuval = "4.214563820932875";
		
		metricName = "MemoryUtilization[1m]{resourceId = \"ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva\"}.mean()";
		resp= executeQuery(compartment, namespace, metricName);
		String memval = "3.850824792171347";
		
		metricName = "NetworksBytesIn[1m]{resourceId = \"ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva\"}.rate()";
		resp= executeQuery(compartment, namespace, metricName);
		String netinval = "2126.26";
		
		metricName = "NetworksBytesOut[1m]{resourceId = \"ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva\"}.rate()";
		resp= executeQuery(compartment, namespace, metricName);
		String netoutval = "1966.16";
		
		return "{\"timestamp\": \"" +ttime+ "\", \"host\": \""+host+"\", \"region\": \""+region+"\", " +
				"\"availabilityDomain\": \""+availabilityDomain+"\", "+
			    "\"faultDomain\": \""+faultDomain+"\", \"shape\": \""+shape+"\", "+
			    "\"cpu_mean\": "+cpuval+", \"mem_mean\": "+memval+", \"net_in\": "+
			    netinval+", \"net_out\": "+netoutval+" }";
    }


    
    public ListMetricsResponse executeQuery(String compartment, String namespace, String metricName){
    	ListMetricsResponse response = null;
    	try {
	        final ListMetricsRequest request =
	                ListMetricsRequest.builder()
	                        .compartmentId(compartment)
	                        .listMetricsDetails(
	                                ListMetricsDetails.builder()
	                                        .namespace(namespace)
	                                        .name(metricName)
	                                        .build())
	                        .build();

	        System.out.printf("Request constructed:\n%s\n\n", request.getListMetricsDetails());
	
	        System.out.println("Sending...");
	        response = monitoringClient.listMetrics(request);
	        
    	} catch (Throwable e) {
          System.err.println("Error fetching Metrics Request: " + e.getMessage());
    	}
    	return response;
    }
    

}


crontab/oci_metrics_vrx.sh
#!/bin/bash
#************************************************************************
#
#   oci_metrics_vrx.sh - get OCI Metrics
#   metadata information into JSON files.
#
#   Copyright 2019  Olivier Perard 
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
#************************************************************************
# Available at: https://github.com/operard/madhack_vrx/blob/master/crontab/
# Created on: 03/11/2019 by Olivier Perard
# Version 1.00
#************************************************************************
set -e

# Define paths for oci-cli and jq or put them on $PATH. Don't use relative PATHs in the variables below.
v_oci="oci"
v_jq="jq"

cpu=$(/home/ubuntu/bin/oci monitoring metric-data summarize-metrics-data --compartment-id ocid1.compartment.oc1..aaaaaaaa4gk5fmtbrfnkfrwwkoef5pmrvxu7dauh52hbisvnhx6rgas5jxja --namespace "oci_computeagent" --query-text "CPUUtilization[1m]{resourceId = "ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva"}.mean()")
		 
mem=$(/home/ubuntu/bin/oci monitoring metric-data summarize-metrics-data --compartment-id ocid1.compartment.oc1..aaaaaaaa4gk5fmtbrfnkfrwwkoef5pmrvxu7dauh52hbisvnhx6rgas5jxja --namespace "oci_computeagent" --query-text "MemoryUtilization[1m]{resourceId = "ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva"}.mean()")

netin=$(/home/ubuntu/bin/oci monitoring metric-data summarize-metrics-data --compartment-id ocid1.compartment.oc1..aaaaaaaa4gk5fmtbrfnkfrwwkoef5pmrvxu7dauh52hbisvnhx6rgas5jxja --namespace "oci_computeagent" --query-text "NetworksBytesIn[1m]{resourceId = "ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva"}.rate()")

netout=$(/home/ubuntu/bin/oci monitoring metric-data summarize-metrics-data --compartment-id ocid1.compartment.oc1..aaaaaaaa4gk5fmtbrfnkfrwwkoef5pmrvxu7dauh52hbisvnhx6rgas5jxja --namespace "oci_computeagent" --query-text "NetworksBytesOut[1m]{resourceId = "ocid1.instance.oc1.eu-frankfurt-1.antheljrqtij3macskkzfsvile75zoka7hmrga3opeuwcycogz5s62hfczva"}.rate()")

host=$(echo $cpu | jq ' .data[0]."dimensions"."resourceDisplayName"')
region=$(echo $cpu | jq ' .data[0]."dimensions"."region"')
availabilityDomain=$(echo $cpu | jq ' .data[0]."dimensions"."availabilityDomain"')
faultDomain=$(echo $cpu | jq ' .data[0]."dimensions"."faultDomain"')
shape=$(echo $cpu | jq ' .data[0]."dimensions"."shape"')

cpulast=$(echo $cpu |  jq ' .data[0]."aggregated-datapoints"' | jq ' .[-1]')
memlast=$(echo $mem |  jq ' .data[0]."aggregated-datapoints"' | jq ' .[-1]')
netinlast=$(echo $netin |  jq ' .data[0]."aggregated-datapoints"' | jq ' .[-1]')
netoutlast=$(echo $netout |  jq ' .data[0]."aggregated-datapoints"' | jq ' .[-1]')

timestamp=$(echo $cpulast | jq ' ."timestamp"')
cpuval=$(echo $cpulast | jq ' ."value"')
memval=$(echo $memlast | jq ' ."value"')
netinval=$(echo $netinlast | jq ' ."value"')
netoutval=$(echo $netoutlast | jq ' ."value"')

#echo "host: $host"
#echo "region: $region"
#echo "availabilityDomain: $availabilityDomain"
#echo "faultDomain: $faultDomain"
#echo "shape: $shape"

#echo "timestamp: $timestamp"
#echo "cpuval: $cpuval"
#echo "memval: $memval"
#echo "netinval: $netinval"
#echo "netoutval: $netoutval"

echo "$timestamp|$host|$region|$availabilityDomain|$faultDomain|$shape|$cpuval|$memval|$netinval|$netoutval" >> /tmp/oci_metrics_vrx.log

curl -k "http://130.61.28.47:18088/services/collector" \
    -H "Authorization: Splunk b81fe15a-0be4-4a1e-82d5-5324888d78f0" \
	-d '{"sourcetype": "_json", "event": {"timestamp": '$timestamp', "host": '$host', "region": '$region', "availabilityDomain": '$availabilityDomain', "faultDomain": '$faultDomain', "shape": '$shape', "cpu_mean": '$cpuval', "mem_mean": '$memval', "net_in": '$netinval', "net_out": '$netoutval' }}'

